[#123] Reduce origination size and provide a simple guide for it

pasqu4le opened the pull request:
## Description

The first two commits of this PR are an attempt to reduce the cost without being to aggressive.
The refactoring that resulted from them (mostly the fist) has been quite successful, but it's still not nearly enough to fit into the limits.

The commit that follows (still a WIP because of missing tests and docs update) solves the origination problem by moving the entrypoint code from the `code` section to lambdas stored in the `storage` and provides a way to generate the parameter needed to load these lambdas into an originated contract

## Related issue(s)

Resolves #123

## :white_check_mark: Checklist for your Pull Request

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

rinn7e commented on ligo/README.md:
```diff
@@ -28,18 +25,60 @@ which will save the compiled contract in `out/baseDAO.tz`.
+parameter in the `out/entrypoints` directory.
+
+For example, to load (or override) an entrypoint into the storage, you can use:
+`tezos-client transfer 0 from alice to myDAO --arg <ep_parameter>`
```
I think we should provide a full example, since it is a bit hard to understand what needs to be done here:


Example: To load the `burn` entrypoint, run this command: 
(the byte can be found in `out/entrypoints/burn.tz`)
```
tezos-client transfer 0 from alice to myDAO --arg (Left (Some (Pair "burn"
                  (Some 0x05020000023f03210316050d0765046e000000062566726f6d5f076504620000000925746f6b656e5f696404620000000725616d6f756e74072f02000000210320074303680100000014454e545259504f494e545f4e4f545f464f554e44032702000001e0034c0317031603210316031603160316034803190325072c0200000000020000001e03200743036c030b0743036801000000094e4f545f41444d494e034203270321031603160317031605700002032105710003031703160342057000020321057100030316057000030317031703420342032103170317034c032105710002031703160570000203210571000303160317057000030316031605700003032105710004057000030321057100040570000303210571000403420329072f020000005b034c0320034c0320074303620000034c03210571000203190325072c020000000203200200000033034c0320074303620000034c03420743036801000000184641325f494e53554646494349454e545f42414c414e4345034203270200000077034c032105710002034c032105710002034b0356072f020000003b057000020320057000020320057000020320034c03420743036801000000184641325f494e53554646494349454e545f42414c414e434503420327020000001c034c0320034c032005700003034c0346057000030570000303420350034c0321057100020317057000020321057100030316031705700003032105710004031603160317031705700003034205700003031603160316034203420342053d036d0342))))
```

pasqu4le replied:
I think that if we provide a full example we should use a very small custom entrypoint, because having a very long sequence of bytes isn't very useful for documentation.

Additionally I think I should add an example of calling this without fake `<ep_parameter>`, but using bash to `cat` an actual entrypoint. How does that sound?

rinn7e replied:
Yeah, I'm good either ways with using a small entrypoint or `cat`.

pasqu4le commented on ligo/README.md:
```diff
@@ -28,18 +25,60 @@ which will save the compiled contract in `out/baseDAO.tz`.
+entrypoint.
+
+To remove a previously loaded entrypoint you can instead use
+`tezos-client transfer 0 from alice to myDAO --arg '(Left (Some (Pair "<ep_name>", None)))'`
```
Yes that sounds like a good idea, I think we can also provide the parameters to load the entrypoints this way too.

pasqu4le commented:
@rinn7e I completed the test implementation and addressed your comments in a new commit.

I'll try to run all the tests in nettest and close #124 as well, but otherwise no more changes are expected for this PR.

@sras I'd also like you to review it as well.

rinn7e reviewed (approved):
Seems good for me!

sras commented on docs/specification-ligo.md:
```diff
@@ -821,19 +860,46 @@ Parameter (in Michelson):
+## Startup mode
 
-BaseDAO allows DAOs to define their own additional entrypoints.
+The contract has a startup mode that can be operated managed and exited at the
```
>that can be operated managed and exited

Something feels weird here, may be a comma is missing, or one of operated/managed is redundant.

sras commented on docs/specification-ligo.md:
```diff
@@ -821,19 +860,46 @@ Parameter (in Michelson):
+The contract has a startup mode that can be operated managed and exited at the
+beginning of its life-cycle, see [Startup Configuration](#startup-configuration).
+
+To fo this the contract `parameter` has:
```
>To fo 

Possible typo

sras commented on ligo/src/proposal.mligo:
```diff
@@ -158,9 +155,13 @@ let check_vote_limit_reached
 
-let vote(votes, config, store : vote_param_permited list * config * storage): return =
+// Note: we require this because we want to put the entrypoint in a lamba and so
+// we need the contract address to be provided (can't use 'SELF_ADDRESS').
```
> we need the contract address to be provided (can't use 'SELF_ADDRESS').

But why does the Edo document say it is allowed in Lambda?

>Instruction SELF_ADDRESS equivalent to SELF; ADDRESS but allowed in lambdas

https://tezos.gitlab.io/protocols/008_edo.html#michelson

pasqu4le replied:
That's correct, but currently LIGO converts it to `SELF; ADDRESS` and so that won't work.

I think we should update this when LIGO is updated after `edo` activation, in the meanwhile I will update this comment.

sras reviewed (approved):
LGTM except for minor things.
